File: Stack/Java/Stack.java
import java.util.NoSuchElementException;

/**
 * Implementation of a classical generic Stack.
 *
 * @author Thomas Lang, 2014
 * @version 0.1
 */
public class Stack<T extends Comparable<T>>{

    /**
     * Representation of a single Node element.
     *
     * @author Thomas Lang, 2014
     * @version 0.1
     */
    private class Node<T extends Comparable<T>>{
        T value;
        Node<T> next;

        /**
         * Constructor, creates a new Node
         * containing the content </code>value</code>.
         *
         * @param  T value
         *                  Value of newly created Node.
         */
        private Node( T value ){
            this.value = value;
            next = null;
        }

        /**
         * String representation of a single Node.
         *
         * @return String representation of a single Node.
         */
        public String toString(){
            return ( "{" + value + "}" );
        }
    }

    Node<T> sp;  // Note: sp stands for "stack pointer" (the pointer to the top element) 
    int size;

    /**
     * Constructor, initializes the stack pointer 
     * to NULL and the size to 0.
     */
    public Stack(){
        sp = null;
        size = 0;
    }

    /**
     * Checks whether a Stack is empty or not.
     *
     * @return <code>true</code> if the Stack is empty, <code>false</code> otherwise.
     */
    public boolean isEmpty(){ return ( sp == null ); }

    /**
     * Pushes a new Node containing the content <code>value</code> 
     * to the Stack.
     *
     * @param  T value
     *                  Value of the new Node.
     */
    public void push( T value ){
        Node<T> newNode = new Node<T>( value );
        if( isEmpty() ){
            sp = newNode;
            size++;
        }
        else{
            newNode.next = sp;
            sp = newNode;
            size++;
        }
    }

    /**
     * Deletes the top element and returns it.
     *
     * @return Top element of the Stack.
     * @throws NoSuchElementException
     */
    public Node<T> pop() throws NoSuchElementException {
        if( isEmpty() )
            throw new NoSuchElementException( "Error! Cannot pop from empty stack." );
        Node<T> top = sp;
        sp = sp.next;
        size--;
        return top;
    }

    /**
     * Returns the top element of the Stack
     * like pop, but NOT DELETING IT.
     *
     * @return Top element of the Stack.
     * @throws NoSuchElementException 
     */
    public Node<T> top() throws NoSuchElementException {
        if( isEmpty() )
            throw new NoSuchElementException( "Error! Cannot top from empty stack." );
        return sp;
    }

    /**
     * String representation of a Stack.
     *
     * @return String representation of a Stack.
     */
    public String toString(){
        if( isEmpty() )
            return "(empty stack)";
        StringBuilder out = new StringBuilder();
        Node<T> front = sp;
        out.append( "---> " + front.value + "\n" );
        while( front.next != null ){
            front = front.next;
            out.append( "     " + front.value + "\n" );
        }
        return out.toString();
    }
}

File: Stack/Python/stack.py
#!/usr/env/python

class Node:
    """Representation of a single Node.
    A node saves a value and a reference
    to its successor."""

    def __init__( self, value ):
        """Constructor for a new Node.
        
        @param  value  The value to save.
        """
        self.value     = value
        self.nextNode  = None

    def getValue( self ):
        """Returns the saved value.
        
        @return The saved value.
        """
        return self.value

    def setValue( self, value ):
        """Sets the saved value to the
        passed parameter.
        
        @param  value  The value to save.
        """
        self.value = value

    def getNext( self ):
        """Receives "next" reference.
        
        @return  The "next" pointer.
        """
        return self.nextNode

    def setNext( self, node ):
        """Sets the "next" reference.
        
        @param  node  The new "next" pointer.
        """
        self.nextNode = node

    def __str__( self ):
        """Returns a nice string representation
        of a single Node-object.
        
        @return String representation of a Node.
        """
        return ("{" + str(self.value) + "}")


class Stack:
    """Implementation of a typical Stack. So a Stack
    is a data-structure that allows access to the most
    top element, the stack pointer, only."""

    def __init__( self ):
        """Constructor for a new, empty Stack."""
        self.sp   = None
        self.size = 0

    def isEmpty( self ):
        """Checks, whether a Stack is empty or not.
        
        @return True if the Stack is empty, False otherwise.
        """
        return (self.sp is None)

    def getSize( self ):
        """Returns the size of the Stack.
        
        @return The size of the Stack.
        """
        return self.size

    def push( self, value ):
        """Pushes a new Node containing the passed parameter
        to the Stack.
        
        @param  value  The value of the new Node.
        """
        node = Node( value )
        if self.isEmpty():
            self.sp = node
        else:
            node.setNext( self.sp )
            self.sp = node 
        self.size += 1

    def pop( self ):
        """Performs the "pop" operation on a Stack,
        deletes top Node and returns it.
        
        @return The top Node.
        """
        if self.isEmpty():
            raise Exception( "Error! Cannot pop from empty Stack." )
        topNode    = self.sp
        self.sp    = self.sp.getNext()
        self.size -= 1
        return topNode

    def top( self ):
        """Returns the top element WITHOUT deleting it."""
        if self.isEmpty():
            raise Exception( "Error! No elements in empty Stack." )
        return self.sp

    def __str__( self ):
        """Returns a String representation of a Stack.
        
        @return A String representation of a Stack.
        """
        out = ""
        if self.isEmpty():
            return "(empty stack)"
        else:
            out += "---> " + str(self.sp) + "\n"
            top = self.sp.getNext()
            while top is not None:
                out += "     " + str(top) + "\n"
                top = top.getNext()
            out += "\n"
        return out



def main():
    """Main-method, does some testing."""
    print( "*** Initializing new Stack ... done." )
    stack = Stack()

    print( "*** Pushing values 7, 42, 24, and 999 to Stack." )
    stack.push( 7   )
    stack.push( 42  )
    stack.push( 24  )
    stack.push( 999 )
    
    print( "   ### Current stack:")
    print( stack )

    print( "   ### Top element: " + str(stack.top()) )
    print( "\n*** Performing pop operation ... done " )
    stack.pop()
    
    print( "   ### Current stack:" )
    print( stack )


if __name__ == '__main__':
    main()
